### PR TITLE
Fix GRPO KL estimator overflow

### LINKED ISSUE #3015: Numerical stability in KL's exp function
I experienced some numerical issues here
https://github.com/huggingface/trl/blob/e3244d2d096ff1e2e248c931d06d39e165e20623/trl/trainer/grpo_trainer.py#L943

My `ref_per_token_logps` and `per_token_logps` differ more during training, leading to a large value in `exp` and getting `inf` on the step.

### PR DESCRIPTION
**Pull request by he-yufeng**
# What does this PR do?

Adds an opt-in `kl_log_ratio_clip` setting for GRPO's KL estimator. When configured, the reference-to-policy log-ratio is clipped before `exp()`, preventing an infinite KL penalty when the policy and reference model have drifted far apart.

The default remains `None`, so existing training behavior is unchanged. The config rejects non-positive values and explicitly rejects the option with Liger until Liger's fused loss supports the same behavior.

Fixes #3015

## Validation

- `python -m pytest tests/test_grpo_trainer.py -k "kl_log_ratio_clip" -q` (4 passed)
- `python -m pytest tests/test_grpo_trainer.py -k "kl_log_ratio_clip or compute_liger_loss_passes_vllm_is_ratio" -q` (4 passed, 1 skipped because Liger is not installed)
- `python -m pytest tests/test_cli.py -k "help_no_type_error and grpo" -q` (1 passed)
- `python -m ruff check trl/trainer/grpo_config.py trl/trainer/grpo_trainer.py tests/test_grpo_trainer.py`
- `git diff --check`

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? See #3015.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed.

---

> [!NOTE]
> **Medium Risk**
> Touches core GRPO loss/KL math across multiple trainers; default None preserves prior behavior, but enabled clipping changes KL penalties during training.
> 
> **Overview**
> Adds optional **`kl_log_ratio_clip`** on `GRPOConfig` (default `None`) so training can cap the reference–policy log-ratio before `exp()` in the KL term, avoiding infinite KL when models diverge. A shared **`_compute_kl`** helper centralizes `exp(log_ratio) - log_ratio - 1` with optional upper clamping; **`GRPOTrainer`** and GRPO-derived trainers (DPPO, GSPO token, PAPO) call it instead of inlined formulas.
> 
> Config validation rejects non-positive clip values and disallows the option with **`use_liger_kernel=True`** until Liger supports the same behavior. Tests cover the helper, end-to-end loss finiteness, and config errors.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 2db26eea35b0979dd12cbd752b90ef3a416080f3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -2547,9 +2547,10 @@ def _compute_loss(self, model, inputs):
         # Compute the KL divergence between the model and the reference model
         if self.beta != 0.0:
             ref_per_token_logps = inputs["ref_per_token_logps"]
-            per_token_kl = (
-                torch.exp(ref_per_token_logps - per_token_logps) - (ref_per_token_logps - per_token_logps) - 1
-            )
+            kl_log_ratio = ref_per_token_logps - per_token_logps
+            if self.args.kl_log_ratio_clip is not None:
+                kl_log_ratio = torch.clamp(kl_log_ratio, max=self.args.kl_log_ratio_clip)
+            per_token_kl = torch.exp(kl_log_ratio) - kl_log_ratio - 1
```
[the remark this answers is not in the document]
Comment (he-yufeng) [reply trl/trainer/grpo_trainer.py]: Fixed in 2db26eea. I moved the KL calculation into a shared helper and routed the base GRPO path plus the GSPO-token, PAPO, and DPPO overrides through it, so the opt-in clip now applies consistently. Focused clip tests and ruff pass; the broader experimental CPU run still has the repository environment's existing bf16/GPU requirement.

**Comment by qgallouedec** [comment]
This profile belongs to a bot that opens a large number of pull requests. In line with our contribution guidelines, this is not the type of contribution we encourage.
